git-config/src/values.rs

```diff
@@ -146,9 +151,8 @@ pub enum Value<'a> {
     Integer(Integer),
     Color(Color),
     /// If a value does not match from any of the other variants, then this
-    /// variant will be matched. As a result, conversion from a `str`-like item
-    /// will never fail.
-    Other(Cow<'a, [u8]>),
```

Comment:
> @Byron do we really need Other/Bytes provided that we already have bytes from the parser?

I think so, another name for it would be `string` I supposed, as it's any other value. The bytes themselves might be backed in some buffer elsewhere, but the `Value` encapsulates a value portion of it. Maybe that's not the answer you were looking for though, am I missing something?

> Also, I don't see why we need the impl blocks of Value, talking of which, impl<'a> From<&'a [u8]> for Value<'a> is not handling cases like impl From<Vec<u8>> for Value<'_> does

Great catch, those should be the same and unified. One might also ask the question where these are called at all and if all of these constructors are necessary. Since ultimately all of the values contain a `Cow<'_, str/[u8]>` I'd expect the general constructor to also take a `Cow<'_, [u8]>`.

In the interest of avoiding opening up pandoras box, I wonder what would happen of only `From/TryFrom<&[u8]>` was supported. Does some code depend on other constructors,, or is `&[u8]` enough effectively? If not, maybe `Cow<'_, [u8]>` could replace the `&[u8]` as general constructor that works for all the cases.

As an outcome of this PR I'd hope there is only one code-paths to construct a value to reduce duplication.

Reply:
Yeah, like we talked it seems like we don't need Value either, so I'll try removing all that

Reply:
That won't work of course since we need to be able to handle strings :-) So now I renamed other to String, moved it outside of the enum, removed the enum and changed the impl block to work with String

Reply:
> In the interest of avoiding opening up pandoras box, I wonder what would happen of only `From/TryFrom<&[u8]>` was supported. Does some code depend on other constructors,, or is `&[u8]` enough effectively? If not, maybe `Cow<'_, [u8]>` could replace the `&[u8]` as general constructor that works for all the cases.

Didn't try that yet since `Boolean`, `Integer` and `Color` work with Cow, so String is like that as well. But I can try.. Do we not use Cow to be able to modify the value?

Reply:
Tried it and getting this error
```
--> git-config/src/file/git_config.rs:2130:28
     |
2130 |         let value = config.value::<String>("remote", Some("origin"), "url").unwrap();
     |                            ^^^^^ the trait `From<Cow<'_, [u8]>>` is not implemented for `values::String<'_>`
```

the function 
`pub fn value<'lookup, T: TryFrom<Cow<'event, [u8]>>>` requires `Cow` or do you mean to change that as well?

Reply:
I am sorry for leading you astray. My comment was aimed towards having one `TryFrom<Cow<…>>` implementation for these wrapper types so they can work with both shared and owned data. I didn't know how that affects the entire system though.

> Do we not use Cow to be able to modify the value?

Cow is used to also be able to create `String<Cow<'static, [u8]>>` types, which effectively own their data and are thus self-contained. This is the kind of storage that's needed for it to be held in a `Repository` instance. Originally I wanted to have a zero-copy version of loading git config files, but it's not realizable here due to this design choice.

---- 

I think it's ok to not put more time into this particular exploration for now - it's already a step forward to not have duplicated constructors.